src/coreclr/jit/fgopt.cpp

```diff
@@ -3726,47 +3727,163 @@ bool Compiler::fgOptimizeBranch(BasicBlock* bJump)
     return true;
 }
 
-/*****************************************************************************
- *
- *  Function called to optimize switch statements
- */
-
+//-----------------------------------------------------------------------------
+// fgOptimizeSwitchJump: see if a switch has a dominant case, and modify to
+//   check for that case up front (aka switch peeling).
+//
+// Returns:
+//    True if the switch now has an upstream check for the dominant case.
+//
 bool Compiler::fgOptimizeSwitchJumps()
 {
-    bool result = false; // Our return value
-
-#if 0
-    // TODO-CQ: Add switch jump optimizations?
     if (!fgHasSwitch)
+    {
         return false;
+    }
 
-    if (!fgHaveValidEdgeWeights)
-        return false;
+    bool modified = false;
 
-    for (BasicBlock* bSrc = fgFirstBB; bSrc != NULL; bSrc = bSrc->bbNext)
+    for (BasicBlock* block = fgFirstBB; block != NULL; block = block->bbNext)
     {
-        if (bSrc->bbJumpKind == BBJ_SWITCH)
+        if (block->IsLIR())
+        {
+            break;
+        }
+
+        if (block->isRunRarely())
         {
-            unsigned        jumpCnt; jumpCnt = bSrc->bbJumpSwt->bbsCount;
-            BasicBlock**    jumpTab; jumpTab = bSrc->bbJumpSwt->bbsDstTab;
+            continue;
+        }
 
-            do
-            {
-                BasicBlock*   bDst       = *jumpTab;
-                flowList*     edgeToDst  = fgGetPredForBlock(bDst, bSrc);
-                double        outRatio   = (double) edgeToDst->edgeWeightMin()  / (double) bSrc->bbWeight;
+        if (block->bbJumpKind != BBJ_SWITCH)
```

Comment:
Is this ever called for LIR? Check for SWITCH before RunRarely, as it's more likely to be false?

Reply:
Doesn't make sense to call for LIR as switches are lowered there. So changed to an assert.